botbot_plugins/plugins/vote.py

```diff
@@ -0,0 +1,334 @@
+                elif throw_error:
+                    return ERROR_MESSAGES["no_vote_started"]
+                return ""
+            return wrap
+        return decorator
+
+    @listens_to_command("endvote")
+    @_depends_on_votestarted()
+    def endvote(self, line, args):
+        name = json.loads(self.retrieve("name"))
+        changes_since_countvote = json.loads(
+            self.retrieve("changes_since_countvote"))
+        reply = (INFO_MESSAGES["voting_ended_name"].format(name=name)
+                 if name else INFO_MESSAGES["voting_ended"])
+        if changes_since_countvote:
+            reply += "\n" + self._print_votes()
+
+        self.delete("name")
+        self.delete("options")
+        self.delete("votes")
+        self.delete("changes_since_countvote")
+
+        return reply
+
+    @listens_to_command("countvotes")
+    @_depends_on_votestarted()
+    def countvotes(self, line, args):
+        self.store("changes_since_countvote", json.dumps(False))
+        return self._print_votes()
+
+    vote_regex_template = (
+        r"^((?P<vote>[\\" +
+        re.escape("|".join(VOTE_SHORTCUT_TO_NAME.keys())) +
+        r"]{vote_modifier})(?P<option>.+))$")
+
+    @listens_to_regex_command("vote", (vote_regex_template
+                                       .format(vote_modifier="?")))
+    @_depends_on_votestarted()
+    def vote(self, line, vote, option):
+        try:
+            changes_made = self._add_vote(self._parse_vote(line.user, (vote, option)))
+            if changes_made:
+                self.store("changes_since_countvote", json.dumps(True))
+        except InvalidOptionError, e:
+            return unicode(e)
+
+    # This makes the vote symbol (+, -, \) required, since otherwise we can't
+    # tell if this is meant to be a vote or not.
+    @listens_to_all(vote_regex_template.format(vote_modifier=""))
+    @_depends_on_votestarted(throw_error=False)
+    def implicit_vote(self, line, vote, option):
+        try:
+            parsed_vote = self._parse_vote(line.user, (vote, option))
+        except InvalidOptionError:
+            return  # Don't complain, will generate too many false positives
+        changes_made = self._add_vote(parsed_vote)
+        if changes_made:
+            self.store("changes_since_countvote", json.dumps(True))
+
+    @listens_to_command("abstain")
+    @_depends_on_votestarted()
+    def abstain(self, line, args):
+        self.store("changes_since_countvote", json.dumps(True))
+        self._add_vote(self._get_abstain_vote(line.user))
+
+    @listens_to_command("cancelvotes")
+    @_depends_on_votestarted()
+    def cancelvotes(self, line, args):
+        self.store("changes_since_countvote", json.dumps(True))
+        votes = json.loads(self.retrieve("votes"))
+        votes = self._remove_all_votes(line.user, votes)
+        self.store("votes", json.dumps(votes))
+
+    def _parse_vote(self, voter, (vote, option)):
+        original_option = option
```

Comment:
I don't really like this constant reassignment, probably more readable to create a new variable for the processed `option` instead of creating a copy of the original and overwriting it.

Reply:
Fixed this, but still doing `option = boolean_shortform` in cases where the option is empty. It's not useful to keep it empty anyway.